File: src/agent-resources/agent_resources/cli/skill.py
"""CLI for skill-add command."""

from typing import Annotated

import typer

from agent_resources.cli.common import get_destination, parse_resource_ref
from agent_resources.exceptions import (
    ClaudeAddError,
    RepoNotFoundError,
    ResourceExistsError,
    ResourceNotFoundError,
)
from agent_resources.fetcher import ResourceType, fetch_resource

app = typer.Typer(
    add_completion=False,
    help="Add Claude Code skills from GitHub to your project.",
)


@app.command()
def add(
    skill_ref: Annotated[
        str,
        typer.Argument(
            help="Skill to add in format: <username>/<skill-name>",
            metavar="USERNAME/SKILL-NAME",
        ),
    ],
    overwrite: Annotated[
        bool,
        typer.Option(
            "--overwrite",
            help="Overwrite existing skill if it exists.",
        ),
    ] = False,
    global_install: Annotated[
        bool,
        typer.Option(
            "--global",
            "-g",
            help="Install to ~/.claude/ instead of ./.claude/",
        ),
    ] = False,
) -> None:
    """
    Add a skill from a GitHub user's agent-resources repository.

    The skill will be copied to .claude/skills/<skill-name>/ in the current
    directory (or ~/.claude/skills/ with --global).

    Example:
        skill-add kasperjunge/analyze-paper
        skill-add kasperjunge/analyze-paper --global
    """
    try:
        username, skill_name = parse_resource_ref(skill_ref)
    except typer.BadParameter as e:
        typer.echo(f"Error: {e}", err=True)
        raise typer.Exit(1)

    dest = get_destination("skills", global_install)
    scope = "user" if global_install else "project"

    typer.echo(f"Fetching skill '{skill_name}' from {username}/agent-resources...")

    try:
        skill_path = fetch_resource(
            username, skill_name, dest, ResourceType.SKILL, overwrite
        )
        typer.echo(f"Added skill '{skill_name}' to {skill_path} ({scope} scope)")
    except RepoNotFoundError as e:
        typer.echo(f"Error: {e}", err=True)
        raise typer.Exit(1)
    except ResourceNotFoundError as e:
        typer.echo(f"Error: {e}", err=True)
        raise typer.Exit(1)
    except ResourceExistsError as e:
        typer.echo(f"Error: {e}", err=True)
        raise typer.Exit(1)
    except ClaudeAddError as e:
        typer.echo(f"Error: {e}", err=True)
        raise typer.Exit(1)


if __name__ == "__main__":
    app()

File: src/agent-resources-npm/src/cli/skill.ts
/**
 * CLI for add-skill command.
 */

import { Command } from 'commander';
import {
  parseResourceRef,
  getDestination,
  createFetchSpinner,
  printSuccessMessage,
} from './common.js';
import { fetchResource } from '../fetcher.js';
import { ResourceType } from '../types.js';
import {
  ClaudeAddError,
  RepoNotFoundError,
  ResourceExistsError,
  ResourceNotFoundError,
} from '../exceptions.js';

/**
 * Create the add-skill CLI command.
 */
export function createSkillCommand(): Command {
  const program = new Command()
    .name('add-skill')
    .description('Add Claude Code skills from GitHub to your project.')
    .argument(
      '<skill-ref>',
      'Skill to add in format: <username>/<skill-name>'
    )
    .option('--overwrite', 'Overwrite existing skill if it exists.', false)
    .option(
      '-g, --global',
      'Install to ~/.claude/ instead of ./.claude/',
      false
    )
    .action(
      async (
        skillRef: string,
        options: { overwrite: boolean; global: boolean }
      ) => {
        try {
          const [username, skillName] = parseResourceRef(skillRef);
          const dest = getDestination('skills', options.global);

          const spinner = createFetchSpinner();
          spinner.start();

          try {
            await fetchResource(
              username,
              skillName,
              dest,
              ResourceType.SKILL,
              options.overwrite
            );
            spinner.stop();
            printSuccessMessage('skill', skillName, username);
          } catch (error) {
            spinner.stop();
            throw error;
          }
        } catch (error) {
          if (
            error instanceof RepoNotFoundError ||
            error instanceof ResourceNotFoundError ||
            error instanceof ResourceExistsError ||
            error instanceof ClaudeAddError
          ) {
            console.error(`Error: ${error.message}`);
            process.exit(1);
          }
          if (error instanceof Error) {
            console.error(`Error: ${error.message}`);
            process.exit(1);
          }
          throw error;
        }
      }
    );

  return program;
}
